bin/helpers/openstack.sh

```diff
@@ -0,0 +1,357 @@
+  server_disc_format=$(get_server_image_disc_format "${image_id}")
+
+  if [[ "${server_disc_format}" == "${disc_format}" ]]; then
+    echo "true"
+  else
+    echo "false"
+  fi
+}
+
+function is_image_already_uploaded() {
+  local image_name=$1
+  local disc_format=$2
+  local image_id
+  local upload_name
+  upload_name="${image_name}"
+  image_id=$(get_image_id "${upload_name}")
+
+  # If the image exists, return true, otherwise return false
+  if [[ "${image_id}" == "" ]]; then
+    echo "false"
+  else
+    echo "true"
+  fi
+}
+
+function wait_for_image_activation() {
+  local upload_name=$1;
+  local wait_time
+  local max_wait
+  local sleep_time
+  local status
+  wait_time=0
+  max_wait=300
+  sleep_time=3
+  status="nothing"
+
+  # Wait until image status is "active"
+  while [[ "${status}" != "active" && ${wait_time} -lt ${max_wait} ]]; do
+    # allow for image to be ready
+    sleep ${sleep_time}
+    wait_time=$((wait_time + sleep_time))
+    status=$(get_image_status "${upload_name}")
+  done
+
+  echo "${status}"
+}
+
+function get_host_uploaded_image_checksum() {
+  local image_name=$1
+  local disc_format=$2
+  local image_url=$3
+  local upload_name="${image_name}"
+  local upload_filepath="/tmp/${image_name}.${disc_format}"
+  local checksum
+
+  if [[ "${disc_format}" == "qcow2" ]]; then
+    # Download the checksum of the qcow2 image from Kubenow servers
+    curl "${image_url}/${upload_name}.md5" \
+      -o "${upload_filepath}.md5" \
+      --connect-timeout 30 \
+      --max-time 30
+
+    checksum=$(cut -f1 -d ' ' "${upload_filepath}.md5")
+  else
+    # Generate the md5 from the local file which has been uploaded
+    checksum=$(md5sum "${upload_filepath}" | awk '{print $1}')
+  fi
+
+  echo "${checksum}"
+}
+
+function verify_uploaded_image() {
+  local image_name=$1
+  local disc_format=$2
+  local image_url=$3
+  local upload_name="${image_name}"
+  local upload_filepath="/tmp/${image_name}.${disc_format}"
+  local image_list
+  local image_id
+  local image_details
+  local server_checksum
+  local host_checksum
+  image_id=$(get_image_id "${upload_name}")
+  server_checksum=$(get_server_uploaded_image_checksum "${image_id}")
+
+  if [[ ${#server_checksum} != 32 ]]; then
+    echo "No valid server_checksum field on server, skipping checksum verification step"
+  else
+    host_checksum=$(get_host_uploaded_image_checksum "${image_name}" "${disc_format}" "${image_url}")
+
+    if [ "${host_checksum}" != "${server_checksum}" ]; then
+      echo >&2 "Server checksum: ${server_checksum}"
+      echo >&2 "Expected checksum: ${host_checksum}"
+      echo >&2 "The server image checksum does not match expectations"
+      echo >&2 "Something might have failed on file transfer."
+      echo >&2 "Please delete image ${image_name} from Openstack and try again."
+      exit 1
+    else
+      echo "Uploaded image checksum is OK"
+    fi
+  fi
+}
+
+function upload_image() {
+  local image_name=$1
+  local disc_format=$2
```

Comment:
To avoid confusion with image format on image downloaded change "disk_format" to "glance_disk_format"